src/Microsoft.AspNet.Razor/Parser/HtmlMarkupParser.Block.cs

```diff
@@ -662,9 +704,20 @@ private bool StartTag(Stack<Tuple<HtmlSymbol, SourceLocation>> tags, IDisposable
                 tagName = new HtmlSymbol(CurrentLocation, String.Empty, HtmlSymbolType.Unknown);
             }
 
+            // Need to check if we need to back-track.
+            if (bangSymbol != null)
+            {
+                // Need to ensure the Tokenizer is pointing to the correct location.
+                PutCurrentBack();
+                PutBack(bangSymbol);
+                NextToken();
```

Comment:
might need more to the comment above since these three lines look identical to a single `PutCurrentBack()` call.  probably has something to do with setting `PreviousSymbol` correctly (and this pattern happens a few times in the code) but comment would still help.